src/3d/qgstessellator.cpp

```diff
@@ -157,22 +157,91 @@ void QgsTessellator::addPolygon( const QgsPolygonV2 &polygon, float extrusionHei
   polyline.reserve( exterior->numPoints() );
 
   QgsVertexId::VertexType vt;
-  QgsPoint pt, ptPrev;
+  QgsPoint pt, ptFirst;
+
+  QVector3D pNormal(0, 0, 0);
+  int pCount = exterior->numPoints();
+  for (int i = 0; i < pCount - 1; i++)
+  {
+    QgsPoint pt1, pt2;
+    exterior->pointAt(i, pt1, vt);
+    exterior->pointAt((i + 1) % pCount, pt2, vt);
+    ptFirst = pt1;
+
+    pNormal.setX(pNormal.x() + (pt1.y() - pt2.y()) * (pt1.z() + pt2.z()));
+    pNormal.setY(pNormal.y() + (pt1.z() - pt2.z()) * (pt1.x() + pt2.x()));
+    pNormal.setZ(pNormal.z() + (pt1.x() - pt2.x()) * (pt1.y() + pt2.y()));
```

Comment:
This is called Newell's method and is referenced in OpenGL [wiki](https://www.khronos.org/opengl/wiki/Calculating_a_Surface_Normal).

It has been one of the approaches tried and the best solution, after all. I have no proof for how it works, but since I have tested it with big datasets, I think it's quite robust.